fix(tags): search for `@name` inside of tags

m4rch3n1ng opened the pull request:
i noticed this while trying to improve the go and rust tags and noticed, that the go tags had some `@name` queries, that didn't seem to do anything, while the rust did not. when querying for the go syntax the picker showed the whole definition: (or at least the first line of it), while the rust picker (correctly) only displayed the name.

<details>
<summary>screenshots</summary>
<img width="1099" height="696" alt="image" src="https://github.com/user-attachments/assets/a9854167-85db-4b3f-b4ed-fe206b17abb9" />
<img width="1099" height="696" alt="image" src="https://github.com/user-attachments/assets/50315c40-c71f-4d36-b5e1-d13169adbae7" />

the go lsp only shows the name:
<img width="1099" height="696" alt="image" src="https://github.com/user-attachments/assets/8f273367-e05e-44c3-9aa4-e5ca2ca724a5" />
</details>

while the rust syntax picker showed the correct name, when selecting a symbol, only selected the name of the definition is selected, while the lsp symbol picker selects the entire definition.

<details>
<summary>screenshots</summary>
(after selecting the `Thing` struct in the picker):
<img width="1099" height="696" alt="image" src="https://github.com/user-attachments/assets/09cd8d0a-6587-4a16-bf5f-6cdd79ccb5b3" />
<img width="1099" height="696" alt="image" src="https://github.com/user-attachments/assets/681421a3-88b3-4a7d-9336-b4a9bbcdb7c4" />
</details>

this pr fixes that by just trying to find a `@name` in the `definition` capture and falling back to displaying whole match if none is found. i am not quite sure with my implementatio

with this pr i can also update the rust tags, so that the syntax query matches the whole definition, which i did in the second commit.

additional notes: the currently ignored `@name` captures are already used by quite a few grammars, including `go`, `ts`/`js`, `python`, `c#` and `elisp` to name a few.

m4rch3n1ng committed:
fix(tags): search for `@name` by using the `TagQuery` of tree-house

m4rch3n1ng committed:
tags(rust): mark the whole definition and use the `@name` queries

m4rch3n1ng committed:
add some docs for the new `@name` tags capture

the-mikedavis commented on helix-term/src/commands/syntax.rs:
```diff
@@ -127,17 +127,38 @@ fn tags_iter<'a>(
+        let start = text.byte_to_char(mat.node.start_byte() as usize);
+        let end = text.byte_to_char(mat.node.end_byte() as usize);
+
+        let mut inner_iter = syntax.tags(text, loader, mat.node.byte_range());
```
We should never need to invoke a new query within a match of that same query. That's working around tags iterator returned captured nodes rather than entire matches. This probably needs work in tree-house to add an iterator over matches rather than captures.

m4rch3n1ng replied:
yeah i expected as much. not quite sure where to start in tree-house, but i'll see what i can come up with ...

the-mikedavis reviewed (commented):
Looking for a `@name` capture in a match of a `@definition.*` capture is correct (we should aim to follow https://tree-sitter.github.io/tree-sitter/4-code-navigation.html). The range of the symbol should be the entire `@definition.*` capture though. Ideally the `tags.scm` query should capture the entire definition with `@definition.*` and the `@name` should be used for display purposes, for example in the picker.

m4rch3n1ng committed:
[tmp] use my git branch for the tree-house dependency

m4rch3n1ng commented:
updated the implementation here to use https://github.com/helix-editor/tree-house/pull/29 and converted to draft, as i have to use a local dependency on tree-house to make it work.

m4rch3n1ng committed:
tags(c): mark the whole definition and use `@name` queries

the-mikedavis commented:
Superseded by https://github.com/helix-editor/helix/pull/15825
